### PR TITLE
[checkpoint] Clean up selective activation checkpoint and make public

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #127959
* #128545
* __->__ #125795

### bc-breaking for existing users of the private API:
- Existing policy functions must now change their return value to be [CheckpointPolicy](https://github.com/pytorch/pytorch/blob/c0b40ab42e38a208351911496b7153511304f8da/torch/utils/checkpoint.py#L1204-L1230)  Enum instead of bool.
   - To restore previous behavior, return `PREFER_RECOMPUTE` instead of `False` and `{PREFER,MUST}_SAVE` instead of `True` depending whether you prefer the compiler to override your policy.
- Policy function now accepts a `ctx` object instead of `mode` for its first argument. 
   - To restore previous behavior, `mode = "recompute" if ctx.is_recompute else "forward"`. 
- Existing calls to `_pt2_selective_checkpoint_context_fn_gen` must be renamed to `create_selective_checkpoint_contexts `. The way you use the API remains the same. It would've been nice to do something different (not make the user have to use functools.partial?), but this was the easiest to compile (idk if this should actually be a constraint). 

Related doc: https://docs.google.com/document/d/1BKyizkZPdri9mHqdDOLAUpkI7SbbKfLHRFVVpK9ZWqo/edit

Memory considerations:
- As with the existing SAC, cached values are cleared upon first use. 
- We error if the user wishes to backward a second time on a region forwarded with SAC enabled.

In-place:
- We use version counting to enforce that if any cached tensor has been mutated. In-place operations not mutating cached tensors are allowed.
- `allow_cache_entry_mutation=True` can be passed to disable this check (useful in the case of auto AC where the user is cleverly also saves the output of the in-place)

Randomness, views
- Currently in this PR, we don't do anything special for randomness or views, the author of the policy function is expected to handle them properly. (Would it would be beneficial to error? - we either want to save all or recompute all random tensors)

Tensor object preservation
- ~We guarantee that if a tensor does not requires grad, and it is saved, then what you get out is the same tensor object.~ UPDATE: We guarantee that if a tensor is of non-differentiable dtype AND it is not a view, and it is saved, then what you get out is the same tensor object. This is a nice guarantee for nested tensors which care about the object identity of of the offsets tensor.

Policy function 
- Enum values are `{MUST,PREFER}_{SAVE,RECOMPUTE}` (bikeshed welcome). Alternatively there was `{SAVE,RECOMPUTE}_{NON_,}OVERRIDABLE`. The former was preferred bc it seemed clearer that two `MUST` clashing should error, versus it is ambiguous whether two `NON_OVERRIDABLE` being stacked should silently ignore or error.
- The usage of Enum today. There actually is NO API to stack SAC policies today. The only thing the Enum should matter for in the near term is the compiler. The stacking SAC policy would be useful if someone wants to implement something like simple FSDP, but it is not perfect because with a policy of `PREFER_SAVE` you are actually saving more than autograd would save normally (would be fixed with AC v3).
- The number of times we call the policy_fn is something that should be documented as part of public API. We call the policy function for all ops except ~~detach~~ UPDATE :  metadata ops listed in `torch.utils.checkpoint.SAC_IGNORED_OPS`) because these ops may be called a different number of times by AC itself between forward and recompute. 
- The policy function can be a stateful object (we do NOT make separate copies of this object for forward/recompute, the user is expected to handle that via is_recompute see below).
Tensors guaranteed to be the same tensor as-is
- Policy function signature takes ctx object as its first argument. The ctx function is an object encapsulating info that may be useful to the user, it currently only holds "is_recompute". Adding this indirection gives us flexibility to add more attrs later if necessary.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/utils/checkpoint.py [hunk 1]
```diff
@@ -1181,133 +1173,233 @@ def _detach(x):
...
-    def _handle_compile_in_forward_ctx(self, should_not_recompute, func, args, kwargs):
         if func in _ignored_ops:
             return func(*args, **kwargs)
-        if should_not_recompute:
-            fx_traceback.current_meta["recompute"] = 0
-        # NOTE: Here we just store and reuse output of all ops, since in torch.compile mode
-        # we decide and handle recomputation in the partitioner.
+
+        if is_compiling:
+            fx_traceback.current_meta["recompute"] = policy
```
Comment (soulitzer) [inline torch/utils/checkpoint.py:1265]: TODO: update the compiler side to handle this

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -1181,133 +1173,233 @@ def _detach(x):
...
+
+        A policy function that returns ``SAVE_{NON,}_OVERRIDABLE`` every op is
+        NOT equivalent to not using checkpointing. Using such a policy would
+        save additional tensors not limited to ones that are actually needed for
+        gradient computation.
+    """
+    SAVE_NON_OVERRIDABLE = 0
+    SAVE_OVERRIDABLE = 1
+    RECOMPUTE_NON_OVERRIDABLE = 2
+    RECOMPUTE_OVERRIDABLE = 3
```
Comment (soulitzer) [inline torch/utils/checkpoint.py:1246]: Maybe bikeshed on this some more.

File: torch/utils/checkpoint.py [hunk 3]
```diff
@@ -1181,133 +1173,233 @@ def _detach(x):
...
+    else:
+        raise TypeError("policy_fn_or_list must be either a function or a list of ops.")
 
+    def context_fn():
+        storage: Dict[Any, List[Any]] = defaultdict(list)
+        # TODO: when would I need to deepcopy the policy_fn?
+        return (
+            _CachingTorchDispatchMode(policy_fn, storage),
+            _CachedTorchDispatchMode(policy_fn, storage),
+        )
```
Comment (fmassa) [inline torch/utils/checkpoint.py:1401]: The `deepcopy` for the policy function can be useful if the policy is not a function, but a class that holds state.

For example, imagine that you have a policy that saves the second matmul. It could look something like
```python
class Policy:
    def __init__(self):
        self.counter = 0

    def __call__(self, mode, func, *args, **kwargs):
            counter = self.counter
            self.counter += 1
            if counter == 1 and func is torch.ops.aten.mm.default:
                return CheckpointPolicy.SAVE_NON_OVERRIDABLE
            return CheckpointPolicy.RECOMPUTE_OVERRIDABLE
                
```

if we don't perform the deepcopy, the `counter` will be shared between the `Caching` and the `Cached` dispatch modes, which will not behave as expected.

This type of pattern is what I used for the [automatic policy selection in xFormers](https://github.com/facebookresearch/xformers/blob/ea4a5508879d31424be5e9b30747fbd162b778a3/xformers/checkpoint.py#L450-L461), which requires finer-grained control of the operations being used, including which operation is used (first, second, etc).

Maybe there is a better way of doing this (like using the `mode` argument to reset the counter), but the deepcopy was easy enough back then (when I didn't have the `mode` argument in the policy, which btw I don't yet use and I'm not sure it's super needed)

Comment (soulitzer) [reply torch/utils/checkpoint.py:1401]: I see, thanks for the context!

> Maybe there is a better way of doing this (like using the mode argument to reset the counter), but the deepcopy was easy enough back then (when I didn't have the mode argument in the policy, which btw I don't yet use and I'm not sure it's super needed)

I think I prefer this idea actually because it may be possible for some one to wish to share some state during the forward and recompute for whatever reason (I don't know any concrete use cases yet). 

One would need to write policy_fn a little differently though. (Hopefully straighforward if we document that a single instance of policy_fn is shared between forward and recompute)
```python
class Policy:
    def __init__(self):
        self.counter = 0
        self.recompute_counter = 0

    def __call__(self, ctx, func, *args, **kwargs):
            counter = self.recompute_counter if ctx.is_recompute() else self.counter
            counter += 1
            if counter == 1 and func is torch.ops.aten.mm.default:
                return CheckpointPolicy.MUST_SAVE
            return CheckpointPolicy.PREFER_RECOMPUTE
```

**Comment by soulitzer** [comment]
TODO: Add a test specifically for NT + SAC when offset is created within the AC region

File: torch/utils/checkpoint.py [hunk 4]
```diff
@@ -1152,149 +1145,209 @@ def _is_compiling(func, args, kwargs):
...
+        policy = self.policy_fn(SelectiveCheckpointContext(is_recompute=True),
+                                func, *args, **kwargs)
+        is_compiling = _is_compiling(func, args, kwargs)
+
+        if policy in (CheckpointPolicy.MUST_SAVE, CheckpointPolicy.PREFER_SAVE) or is_compiling:
+            storage = self.storage.get(func)
+            if func is None:
+                raise RuntimeError("")
+            if len(storage) == 0:
+                raise RuntimeError("Trying to backward an extra time")
```
Comment (soulitzer) [inline torch/utils/checkpoint.py:1262]: (TODO) Add: "You are only allowed to backward once on any region computed under selective activation checkpoint."

**Comment by Chillee** [comment]
What are the exact semantics of `{MUST, PREFER}_{SAVE, RECOMPUTE}`? 

In particular, in combination with the memory budget API, what is the compiler allowed to recompute? Also, how does stacked selective checkpointing work?

MUST_SAVE + PREFER_RECOMPUTE => MUST_SAVE I'm guessing?

**Review by Chillee — COMMENTED** [review]
Also, how should we combine this with the `memory_budget` API?

**Comment by soulitzer** [comment]
> What are the exact semantics of {MUST, PREFER}_{SAVE, RECOMPUTE}?

Happy to discuss this some more if people have differing opinion, but one idea is:
- inner beats outer when `PREFER` and `PREFER` because whoever acts later likely has the most information/context and thus should have the highest priority. This implies compiler beats `PREFER` because compiler acts later with knowledge of the whole graph.
- `MUST` beats `PREFER` regardless of ordering. 
- `MUST `and `MUST` produces an error (due to contradiction because reading literally, both must be true!)

> In particular, in combination with the memory budget API, what is the compiler allowed to recompute?

Anything is not marked `MUST_SAVE` can selected by compiler recomputed.

> Also, how does stacked selective checkpointing work?

My initial idea here is that there would be a separate API (context manager?) that user enters into during forward. Due to recompute we'd enter into this same context manager during backward. Entering the context pushes a policy onto the stack. The TorchDispatchMode applies each policy in the stack and uses the rules above to decide SAVE/RECOMPUTE or error.

File: torch/_higher_order_ops/wrap.py [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
 from torch._ops import HigherOrderOperator
-from torch.utils.checkpoint import checkpoint, uid
+from torch.utils.checkpoint import checkpoint
+
 import torch._dynamo.config
 
 log = logging.getLogger(__name__)
 
-
+uid = itertools.count(1)
```
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/_higher_order_ops/wrap.py:13]: I see it used in the impl of TagActivationCheckpoint:
```
def tag_nodes(self, gmod):
        unique_graph_id = next(uid)
        for node in gmod.graph.nodes:
            if node.op in ("call_function", "call_method", "call_module"):
                node.meta["recompute"] = unique_graph_id
        return gmod
```

File: docs/source/checkpoint.rst [hunk 6]
```diff
@@ -35,3 +35,6 @@ torch.utils.checkpoint
 .. autofunction:: checkpoint
 .. autofunction:: checkpoint_sequential
 .. autofunction:: set_checkpoint_debug_enabled
+.. autoclass:: CheckpointPolicy
+.. autoclass:: SelectiveCheckpointContext
+.. autofunction:: gen_selective_checkpoint_context_fn
```
Comment (Chillee) [inline docs/source/checkpoint.rst:40]: Bikeshedding: What do you think of the name `create_sac_context_fn`. I don't quite like the prefix `gen` - it's somewhat ambiguous to me.

**Review by Chillee — APPROVED** [review]
LGTM!

File: docs/source/checkpoint.rst (hunk 6 above)
Comment (soulitzer) [reply docs/source/checkpoint.rst:40]: Agreed on `gen`. Tentatively updating to`create_selective_checkpoint_contexts`:
- `sac` -> `selective_checkpoint` to avoid acronyms (we don't use AC anywhere)
- `context_fn` -> `contexts` to better reflect that we're actually returning 2 contexts.

File: torch/utils/checkpoint.py [hunk 7]
```diff
@@ -1153,149 +1146,226 @@ def _is_compiling(func, args, kwargs):
...
-            return self._handle_compile_in_recompute_ctx(should_not_recompute, func, args, kwargs)
+
+        kwargs = {} if kwargs is None else kwargs
+        policy = self.policy_fn(SelectiveCheckpointContext(is_recompute=True),
+                                func, *args, **kwargs)
+        is_compiling = _is_compiling(func, args, kwargs)
+
+        if policy in (CheckpointPolicy.MUST_SAVE, CheckpointPolicy.PREFER_SAVE) or is_compiling:
+            storage = self.storage.get(func)
+            if func is None:
```
Comment (fmassa) [inline torch/utils/checkpoint.py:1272]: Shouldn't this be `if storage is None` ?

File: torch/utils/checkpoint.py [hunk 8]
```diff
@@ -1153,149 +1146,226 @@ def _is_compiling(func, args, kwargs):
...
+        >>> out = torch.utils.checkpoint.checkpoint(
+        >>>     fn, x, y,
+        >>>     use_reentrant=False,
+        >>>     context_fn=context_fn,
+        >>> )
     """
-    storage: Dict[Any, List[Any]] = defaultdict(list)
-    return _CachingTorchDispatchMode(policy_fn, storage), _CachedTorchDispatchMode(policy_fn, storage)
+    # NB: If grad_mode is disabled, checkpoint would not run forward under
+    #     context_fn anyway, so proceed as usual.
```
Comment (fmassa) [inline torch/utils/checkpoint.py:1341]: Ah nice, this should be something newer then that wasn't present in the first version I did a while back.

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

I think there is only a minor check issue that would be good to be addressed before merge, otherwise good to merge!

File: torch/utils/checkpoint.py (hunk 7 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py:1272]: whoops, nice catch

**Comment by awgu** [comment]
> Existing calls to _pt2_selective_checkpoint_context_fn_gen must be renamed to gen_selective_checkpoint_context_fn. The way you use the API remains the same. It would've been nice to do something different (not make the user have to use functools.partial?), but this was the easiest to compile (idk if this should actually be a constraint).

should we update the PR description to say `create_selective_checkpoint_contexts` instead of `gen_selective_checkpoint_context_fn`? :)

**Comment by soulitzer** [comment]
Whoops, updated!

**Comment by Chillee** [comment]
I think @wanchaol said he fixed it?

**Comment by wanchaol** [comment]
> I think @wanchaol said he fixed it?

@Chillee fixed API issues and eager works, but compile crashed (we missed a e2e compile + SAC integration test in CI).. discussed offline with @soulitzer and it seems like there're couple of compile issues, we decided to first revert this PR and after fixing those bugs we can reland it.

**Comment by soulitzer** [comment]
Locally SAC + compile works on TorchTitan llama3_8b now with the changes just committed. TODO - This PR may need more testing to cover that case.

File: torch/utils/checkpoint.py [hunk 9]
```diff
@@ -1153,149 +1147,246 @@ def _is_compiling(func, args, kwargs):
...
+                )
+        return self.val
+
+
+def _maybe_detach(x, any_ret_has_alias_info):
+    # We detach for two separate reasons:
+    # - For view ops, detach clears the autograd meta so that when the
+    #   tensor is returned from CachedDispatchMode, as_view does not
+    #   complain about autograd meta already existing.
+    # - For ops that have grad_fn, we want to detach from the graph to
```
Comment (soulitzer) [inline torch/utils/checkpoint.py:1171]:
```suggestion
    # - For tensors that have grad_fn, we want to detach from the graph to
```
